### PR TITLE
SONARJS-708 Consider "instanceof" expressions

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* CRITICAL 1 critical
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: its/ruling/src/test/expected/javascript-S2583.json [hunk 1]
```diff
@@ -594,8 +605,7 @@
 410,
 ],
 'project:google-closure-library/base.js':[
-701,
-723,
+661,
```
Comment (pynicolas) [inline its/ruling/src/test/expected/javascript-S2583.json]: @vilchik-elena This looks like a false positive, for example when `value` is created with `Object.create(null)`.

File: javascript-frontend/src/main/java/org/sonar/javascript/se/sv/InstanceOfSymbolicValue.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    Constraint constraintForObject = null;
+    if (constructorValue instanceof BuiltInObjectSymbolicValue) {
+      constraintForObject = ((BuiltInObjectSymbolicValue) constructorValue).type().constraint();
+    }
+
+    if (constraint.isStricterOrEqualTo(Constraint.TRUTHY)) {
+      newProgramState = state.constrain(objectValue, constraintForObject == null ? Constraint.NOT_NULLY : constraintForObject);
+
+    } else {
+      constraintForObject = constraintForObject == null || constraintForObject == Constraint.OBJECT ? Constraint.ANY_VALUE : constraintForObject.not();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/sv/InstanceOfSymbolicValue.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Pull request merged**